cli/file_watcher.rs

```diff
@@ -74,180 +77,127 @@ async fn error_handler(watch_future: FileWatcherFuture<Result<(), AnyError>>) {
   }
 }
 
-/// This function adds watcher functionality to subcommands like `fmt` or `lint`.
-/// The difference from [`watch_func_with_module_resolution`] is that this doesn't depend on
-/// [`ModuleGraph`].
-///
-/// - `target_resolver` is used for resolving file paths to be watched at every restarting of the watcher. The
-/// return value of this closure will then be passed to `operation` as an argument.
-///
-/// - `operation` is the actual operation we want to run every time the watcher detects file
-/// changes. For example, in the case where we would like to apply `fmt`, then `operation` would
-/// have the logic for it like calling `format_source_files`.
-///
-/// - `job_name` is just used for printing watcher status to terminal.
-///
-/// Note that the watcher will stop working if `target_resolver` fails at some point.
-///
-/// [`ModuleGraph`]: crate::module_graph::Graph
-pub async fn watch_func<F, G>(
-  target_resolver: F,
-  operation: G,
-  job_name: &str,
-) -> Result<(), AnyError>
+pub enum ResolutionResult<T> {
+  Restart {
+    paths_to_watch: Vec<PathBuf>,
+    result: Result<T, AnyError>,
+  },
+  Ignore,
+}
+
+async fn next_restart<F, T: Clone>(
+  resolver: &mut F,
+  debounce: &mut Pin<&mut Debounce>,
+  initial: bool,
+) -> (Vec<PathBuf>, Result<T, AnyError>)
 where
-  F: Fn() -> Result<Vec<PathBuf>, AnyError>,
-  G: Fn(Vec<PathBuf>) -> FileWatcherFuture<Result<(), AnyError>>,
+  F: FnMut(Option<Vec<PathBuf>>) -> FileWatcherFuture<ResolutionResult<T>>,
 {
-  let debounce = Debounce::new();
-  pin!(debounce);
-
+  let mut changed = if initial { None } else { debounce.next().await };
   loop {
-    let paths = target_resolver()?;
-    let _watcher = new_watcher(&paths, &debounce)?;
-    let func = error_handler(operation(paths));
-    let mut is_file_changed = false;
-    select! {
-      _ = debounce.next() => {
-        is_file_changed = true;
-        info!(
-          "{} File change detected! Restarting!",
-          colors::intense_blue("Watcher"),
-        );
-      },
-      _ = func => {},
-    };
-
-    if !is_file_changed {
-      info!(
-        "{} {} finished! Restarting on file change...",
-        colors::intense_blue("Watcher"),
-        job_name,
-      );
-      debounce.next().await;
-      info!(
-        "{} File change detected! Restarting!",
-        colors::intense_blue("Watcher"),
-      );
+    let initial = changed.is_none();
+    match resolver(changed).await {
+      ResolutionResult::Ignore => {
+        debug!("File change ignored")
+      }
+      ResolutionResult::Restart {
+        paths_to_watch,
+        result,
+      } => {
+        if !initial {
+          info!(
+            "{} File change detected! Restarting!",
+            colors::intense_blue("Watcher"),
+          );
+        }
+        break (paths_to_watch, result);
```

Comment:
Here, is this `break` exactly equivalent to `return`? If so, personally I prefer `return` to `break` for readability.